oak_functions/loader/src/server.rs

```diff
@@ -332,6 +335,45 @@ impl WasmState {
         Ok(())
     }
 
+    pub fn invoke_extension_with_handle(
+        &mut self,
+        handle: AbiExtensionHandle,
+        args: wasmi::RuntimeArgs,
+    ) -> Result<Option<wasmi::RuntimeValue>, wasmi::Trap> {
+        let handle: ExtensionHandle =
+            ExtensionHandle::from_i32(handle).expect("Fail to parse handle.");
+
+        // Quick solution following impelementation of invoking an extension in `invoke_index`.
+
+        // First, we get all extensions from WasmState.
+        let mut extensions_indices = self
+            .extensions_indices
+            .take()
+            .expect("no extensions_indices is set");
+
+        // Then we get the extension which has the given handle by looking at the values of the
+        // extension_indices.
+        let extension = extensions_indices
+            .iter_mut()
```

Comment:
Why is a mutable iterator needed? It seems weird that this requires taking the extension indicices and then putting them back when nothing seems to be actually mutated.

Reply:
Because `extension.invoke(self, ...)` in line 369 needs a mutable `WasmState` and `extension_indices` are part of `WasmState`.

With  #2664 we can simplify this.

Reply:
Ah, I see. I guess part of the weirdness it that an `Option` is used rather than a `RefCell` to implement interior mutability, but this is preexisting so can be fixed later as part of #2664 and other refactoring.